optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,456 @@
+            distributions.append(param_distribution)
+            param_names.append(param_name)
+            for parent_index, trial in enumerate(parents):
+                param = trans.transform({param_name: trial.params[param_name]})[0]
+                parents_not_categorical_params[parent_index].append(param)
+
+        xs = np.array(parents_not_categorical_params)
+
+        if crossover_name == "uniform":
+            params_array = _uniform(xs[0], xs[1], rng, swapping_prob)
+        elif crossover_name == "blxalpha":
+            alpha = 0.5
+            params_array = _blxalpha(xs[0], xs[1], rng, alpha)
+        elif crossover_name == "sbx":
+            if len(study.directions) == 1:
+                eta = 2
+            else:
+                eta = 20
+            params_array = _sbx(xs[0], xs[1], rng, distributions, eta)
+        elif crossover_name == "vsbx":
+            if len(study.directions) == 1:
+                eta = 2
+            else:
+                eta = 20
+            params_array = _vsbx(xs[0], xs[1], rng, eta)
+        elif crossover_name == "undx":
+            sigma_xi = 0.5
+            sigma_eta = 0.35 / np.sqrt(len(xs[0]))
+            params_array = _undx(xs[0], xs[1], xs[2], rng, sigma_xi, sigma_eta)
+        elif crossover_name == "undxm":
+            _m = len(xs) - 2
+            _n = len(xs[0])
+            assert _n - _m > 0
+
+            sigma_xi = 1 / np.sqrt(_m)
+            sigma_eta = (
+                0.35
+                * np.sqrt(_m + 1)
+                * np.sqrt(3)
+                / np.sqrt(_n - _m)
+                / np.sqrt(_m + 2)
+                / np.sqrt(2)
+            )
+            params_array = _undxm(xs, rng, sigma_xi, sigma_eta)
+        elif crossover_name == "spx":
+            epsilon = np.sqrt(len(xs[0]) + 2)
+            params_array = _spx(xs, rng, epsilon)
+        else:
+            assert False
+
+        _params = [
+            trans.untransform(np.array([param])) for trans, param in zip(transes, params_array)
+        ]
+        for param in _params:
+            for param_name in param.keys():
+                child[param_name] = param[param_name]
+
+        if _is_contained(child, search_space):
+            break
+
+    return child
+
+
+def _selection(
+    crossover_name: str,
+    study: Study,
+    parent_population: Sequence[FrozenTrial],
+    rng: np.random.RandomState,
+    dominates: Callable[[FrozenTrial, FrozenTrial, Sequence[StudyDirection]], bool],
+) -> List[FrozenTrial]:
+    if crossover_name in ["uniform", "blxalpha", "sbx", "vsbx"]:
+        n_select = 2
+    elif crossover_name in ["undx", "spx"]:
+        n_select = 3
+    elif crossover_name in ["undxm"]:
+        n_select = 4
+    else:
+        assert False
+    if len(parent_population) < n_select:
+        raise ValueError(
+            f"Using {crossover_name},"
+            f"the population size should be greater than or equal to {n_select}."
+        )
```

Comment:
Is it documented that the `ValueError` occurs with an invalid `parent_population`?  In Optuna, all possible errors should be documented like [here](https://github.com/optuna/optuna/blob/master/optuna/samplers/_tpe/sampler.py#L214)

Reply:
On a different note, do we need to raise the `ValueError` here? This function is called internally, so it seems to be sufficient that we assert the condition like `assert len(parent_population) >= n_select`.

Reply:
Thank you for your comment.
I have a question about here.
How can `assert` and `ValueError` be used in different ways?

`ValuError` : notifies the user of an expected bug and prompts him to change the hyperparameters.
`assert` : detects a bug that the developer did not anticipate (the developer assumes that the assert will not cause the program to terminate)

I used ValueError here because I thought it was above.